File: src/detectorpy/detector.py
from numpy import ndarray, array
from cv2 import cvtColor, CascadeClassifier, COLOR_BGR2GRAY


DEFAULT_THRESHOLD = 100
DEFAULT_SCALE_FACTOR = 1.1
DEFAULT_MIN_NEIGHBORS = 4


class Detector:

    def __init__(self,
        classifier_path: str,
        threshold: int = DEFAULT_THRESHOLD,
        scale_factor: float = DEFAULT_SCALE_FACTOR,
        min_neighbors: int = DEFAULT_MIN_NEIGHBORS
    ):
        self.classifier = CascadeClassifier(classifier_path)
        self.threshold = threshold
        self.scale_factor = scale_factor
        self.min_neighbors = min_neighbors

    def check_frame(self, frame: ndarray):
        """
        Takes a frame in the BGR color space in the form
        of a numpy array and uses self.classifier to
        determine if a proximity warning is required.
        """

        # convert bgr frame to grayscale
        gray_frame = cvtColor(frame, COLOR_BGR2GRAY)

        # locate objects in frame
        objects = self.classifier.detectMultiScale(gray_frame, self.scale_factor, self.min_neighbors)

        # if no objects are found, return an empty numpy array and no warning
        if not len(objects):
            return array([]), False

        # find the closest object
        closest = max(objects, key=lambda x: x[1] + x[3])

        # check if a proximity warning should be given
        warning = closest[1] + closest[3] > frame.shape[0] - self.threshold

        # return bounding box of closest object and warning decision
        return closest, warning

File: src/detectorcpp/detector.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include "detector.hpp"


Detector::Detector(std::string classifier_path, int t, int s, int m) {
    threshold = t;
    scale_factor = s;
    min_neighbors = m;
    classifier.load(classifier_path);
}


bool Detector::check_frame(cv::Mat frame, int &n, cv::Rect &closest) {

    // convert frame from bgr to grayscale colorspace
    cv::Mat gray_frame;
    cv::cvtColor(frame, gray_frame, cv::COLOR_BGR2GRAY);

    // locate objects in frame
    std::vector<cv::Rect> objects;
    classifier.detectMultiScale(gray_frame, objects);
    n = objects.size();

    // if no objects have been found, return no warning and leave closest unchanged
    if (!n) {
        return false;
    }

    // find "closest" rectangle
    cv::Rect closest_obj;
    for (int i=0; i<n; i++) {
        if (objects[i].y + objects[i].height > closest_obj.y + closest_obj.height) {
            closest_obj = objects[i];
        }
    }
    closest = closest_obj;

    // check if a proximity warning should be given
    bool warning = closest_obj.y + closest_obj.height > frame.size().height - threshold;

    return warning;

}
